### PR TITLE
[PULL REQUEST] Add updates to improve numerical stability

### PR DESCRIPTION
**Pull request by yantosca**
## Overview
This PR addresses several issues in HEMCO that were causing numerical differences in output:

  1. Now use the appropriate kind parameters to avoid precision loss in numerical constants, in many routines and modules:
     -  Use `1.0_hp` or `1.0d0` instead of `1.0`, etc.
     
  2. Avoid testing for floating point equality, in many routines and modules:
     -  Use `IF ( .not. ABS( x ) > 0.0d0 )` instead of `IF ( x == 0.0d0 )`
    
  3. Use the same values for physical constants as defined in GEOS-Chem's `physconstants.F90` module
  
  4. Make sure to zero loop variables that are declared as `!$OMP PRIVATE` at the top of the parallel loop.  This will prevent junk variables from causing randomization in results.
  
  5. Use `!$OMP  COLLAPSE` to improve the efficiency of parallel loops where expedient.
  
  6. in `src/Core/hco_config_mod.F90`, abstract the code that processes time cycle information into a subroutine that is called from two locations rather than duplicating the code twice.  Also make sure that `Dta%Discontinuous = .FALSE.`.
     - This led to a problem where the wrong time slice of MODIS LAI data was being fetched, because HEMCO thought it was dealing with discontinuously-spaced data in the interpolation routines.
     
  7.  In routine `GET_CURRENT_EMISSIONS`, use an `!$OMP REDUCTION` to make sure that non-private variables `tmpLL` and `nnLL` are summed into properly.
  
  8. In `src/Extensions/hcox_iodine_mod.F90`:
     - Parallelized the main do loop
     - Added more clear comments (especially for unused or debug code)
     - Consolidated IF blocks
     - Cosmetic changes
   
  10. In 'src/Extensions/hcox_paranox_mod.F90`:
      - Zero `Inst%ShipNO` in routine `Evolve_Plume` before it is populated
      - Added a more robust test for zero (as described above)
      - Cosmetic changes

## Related Issues
See: #132, #133, https://github.com/geoschem/geos-chem/issues/1157, https://github.com/geoschem/geos-chem/issues/1103

## Related Pull Requests
There is a corresponding PR: https://github.com/geoschem/geos-chem/pull/1169. Both of these PRs should be merged at the same time.

### CODE REVIEW

**Review by msulprizio — COMMENTED** [review]
These changes look OK, but I just had a few questions as noted below.

File: src/Extensions/hcox_iodine_mod.F90 [hunk 1]
```diff
@@ -162,115 +162,122 @@ SUBROUTINE HCOX_Iodine_Run( ExtState, HcoState, RC )
...
-!       ! SST dependence of iodide - Macdonald et al. 2014
+!%%%% Comment out unused code (bmy, 09 Mar 2022)
+!%%%#if defined( MODEL_GEOS )
+!%%%!       ! Empirical SST scaling factor (jaegle 5/11/11)
+!%%%!       SCALE = 0.329d0 + 0.0904d0*SST - &
+!%%%!               0.00717d0*SST**2d0 + 0.000207d0*SST**3d0
+!%%%#endif
+!%%%!
+!%%%!       ! SST dependence of iodide - Chance et al. 2014
+!%%%!       IODIDE = ( (0.225d0 * ( (SST)**2d0) )  + 19d0 )  / 1d9
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Extensions/hcox_iodine_mod.F90:219]: @msulprizio, they were commented out within the MODEL_GEOS block but the MODEL_GEOS block itself was not commented out.  So I decided to leave it there but just comment everything out and note it as such.

File: src/Core/hco_calc_mod.F90 [hunk 2]
```diff
@@ -3370,14 +3384,14 @@ SUBROUTINE Get_Current_Emissions_Adj( HcoState, BaseDct,   &
...
 !             write(lun,*) 'Updt (IX,IY,L1): ', OUTARR_3D(IX,IY,1)
           endif
 
        ENDDO !I
        ENDDO !J
-!$OMP END PARALLEL DO
+!!$OMP END PARALLEL DO
 
        ! error check
        IF ( ERROR > 0 ) THEN
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_calc_mod.F90:3397]: I've made the last one ERROR==4.  I think that is right.  That pre-dated my modifications.

File: src/Core/hco_calc_mod.F90 [hunk 3]
```diff
@@ -3191,23 +3205,23 @@ SUBROUTINE Get_Current_Emissions_Adj( HcoState, BaseDct,   &
...
 
        ! Loop over all latitudes and longitudes
-!$OMP PARALLEL DO                                                      &
-!$OMP DEFAULT( SHARED )                                                &
-!$OMP PRIVATE( I, J, tIdx, TMPVAL, L, LowLL, UppLL, tmpLL, MaskScale ) & 
-!$OMP SCHEDULE( DYNAMIC )
+!!$OMP PARALLEL DO                                                      &
+!!$OMP DEFAULT( SHARED )                                                &
+!!$OMP PRIVATE( I, J, tIdx, TMPVAL, L, LowLL, UppLL, tmpLL, MaskScale ) &
+!!$OMP SCHEDULE( DYNAMIC )
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_calc_mod.F90]: Somehow those updates got clobbered.  Now restored in commits 45b08c299ae2acf3f6980783b382182f5e569bcc and 13258d1e0b86a313e669e2cc7bafa693f514dd4b.

File: src/Core/hco_calc_mod.F90 [hunk 4]
```diff
@@ -3025,13 +3035,17 @@ SUBROUTINE Get_Current_Emissions_Adj( HcoState, BaseDct,   &
...
 
     ! Loop over all latitudes and longitudes
-!$OMP PARALLEL DO                                                      &
-!$OMP DEFAULT( SHARED )                                                &
-!$OMP PRIVATE( I, J, L, tIdx, TMPVAL, DilFact, LowLL, UppLL          ) & 
-!$OMP SCHEDULE( DYNAMIC )
+!!$OMP PARALLEL DO                                                      &
+!!$OMP DEFAULT( SHARED )                                                &
+!!$OMP PRIVATE( I, J, L, tIdx, TMPVAL, DilFact, LowLL, UppLL          ) &
+!!$OMP SCHEDULE( DYNAMIC )
```
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_calc_mod.F90]: Somehow those updates got clobbered.  Now restored in commits 45b08c299ae2acf3f6980783b382182f5e569bcc and 13258d1e0b86a313e669e2cc7bafa693f514dd4b.

File: src/Core/hco_calc_mod.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (yantosca) [reply src/Core/hco_calc_mod.F90]: Now fixed in commit 45b08c299ae2acf3f6980783b382182f5e569bcc

[the remark this answers is not in the document]
Comment (msulprizio) [reply src/Core/hco_calc_mod.F90]: Yes, I did notice those were in the code before your updates. Thanks for fixing those lines too!

**Comment by yantosca** [comment]
Remaining parallelization issues in GEOS-Chem Classic simulations have been solved in PR https://github.com/geoschem/geos-chem/pull/1190.
